keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -1066,8 +1069,191 @@ the remaining requests in that queue start getting faster service.  In
+exposing defining the width of requests via FlowSchema API), we want to start
+with something relatively simple to first get operational experience with it
+before investing into sophisticated algorithms or exposing a knob to users.
+
+In order to determine the function that will be approximating the `width` of
+a request, we should first estimate how expensive a particular request is.
+And we need to think about both dimensions that we're trying to protect from
+overloading (CPU and RAM) and how many concurrency units a request can actually
+consume.
+
+Let's start with CPU.  The total cost of processing a LIST request should be
+proportional to the number of processed objects.  However, given that in
+practice processing a single request isn't parallelized (and the fact that
+we generally scale the number of total concurrency units linearly with amount
+of available resources), a single request should consume no more than A
+concurrency units.  Fortunately that all compiles together because the
+`processing latency` of the LIST request is actually proportional to the
+number of processed objects, so the cost of the request (defined above as
+`<width> x <processing latency>` really is proportaional to the number of
+processed objects as expected.
+
+For RAM the situation is actually different.  In order to process a LIST
+request we actually store all objects that we process in memory. Given that
+memory is uncompressable resource, we effectively need to reserve all that
+memory for the whole time of processing that request.  That suggests that
+the `width` for the request from the RAM perspective should be proportional
+to the number of processed items.
+
+So what we get is that:
+```
+  width_cpu(N) = min(A, B * N)
+  width_ram(N) = D * N
+```
+where N is the number of items a given LIST request is processing.
+
+The question is how to combine them to a single number.  While the main goal
+is to stay on the safe side and protect from the overload, we also want to
+maxiumize the utilization of the available concurrency units.
+Fortunately, when we normalize CPU and RAM to percentage of available capacity,
+it appears that almost all requests are much more cpu-intensive.  Assuming
+4GB:1CPU ratio and 10kB average object and the fact that processing larger
+number of objects can utilize exactly 1 core, that means that we need to
+process 400.000 objects to make the memory cost higher.
+This means, that we can afford the potential minor efficiency that extremely
+large requests would cause and just approximate it by protecting every resource
+independently, which translates to the following function:
+```
+  width(n) = max(min(A, B * N), D * N)
+```
+We're going to better tune the function based on experiments, but based on the
+above back-of-envelope calculations showing that memory should almost never be
+a limiting factor we will apprximate the width simply with:
+```
+width_approx(n) = min(A, ceil(N / E)), where E = 1 / B
```

Comment:
1. On a machine with 16 CPUs and a concurrency limit of 600 earlier, would we set A=38 and E=10666?
2. Wouldn't we want the `ceil` on the outside, to ensure the approximation is always an integer?

Reply:
As discussed during the meeting:
- the min is used to simulate that processing a single request isn't parallelized, so we won't consume all available capacity for large enough request.

So min is not for solving two-dimensions problem - if you look 3 lines above you have:
"... that memory should almost never be a limiting factor we will apprximate the width simply with"

this shows that this is purely for cpu (so single dimensional problem).


> On a machine with 16 CPUs and a concurrency limit of 600 earlier, would we set A=38 and E=10666?

As mentioned in other comments - I'm not going to comment numbers, because I don't believe the default numbers for that are what the reasonable operators will be using.